rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1132 @@
+  }
+}
+
+rmw_dds_common::msg::Gid
+gid_msg_from_string(const std::string & str)
+{
+  rmw_dds_common::msg::Gid gid;
+  gid.data = {0};
+  EXPECT_LT(str.size(), RMW_GID_STORAGE_SIZE);
+  std::strncpy(
+    reinterpret_cast<char *>(gid.data.data()),
+    str.c_str(),
+    RMW_GID_STORAGE_SIZE);
+  return gid;
+}
+
+struct NodeInfo
+{
+  std::string gid;
+  std::string namespace_;
+  std::string name;
+};
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+add_nodes(
+  GraphCache & graph_cache,
+  const std::vector<NodeInfo> & node_info)
+{
+  rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+  for (const auto & elem : node_info) {
+    msg = graph_cache.add_node(
+      gid_from_string(elem.gid),
+      elem.name,
+      elem.namespace_);
+  }
+  return msg;
+}
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+remove_nodes(
+  GraphCache & graph_cache,
+  const std::vector<NodeInfo> & node_info)
+{
+  rmw_dds_common::msg::ParticipantEntitiesInfo msg;
+  for (const auto & elem : node_info) {
+    msg = graph_cache.remove_node(
+      gid_from_string(elem.gid),
+      elem.name,
+      elem.namespace_);
+  }
+  return msg;
+}
+
+struct NodeEntitiesInfo
+{
+  std::string namespace_;
+  std::string name;
+  std::vector<std::string> reader_gids;
+  std::vector<std::string> writer_gids;
+};
+
+struct ParticipantEntitiesInfo
+{
+  std::string gid;
+  std::vector<NodeEntitiesInfo> node_entities_info_seq;
+};
+
+void
+check_participant_entities_msg(
+  const rmw_dds_common::msg::ParticipantEntitiesInfo & msg,
+  const ParticipantEntitiesInfo & expected)
+{
+  EXPECT_EQ(msg.gid, gid_msg_from_string(expected.gid));
+  ASSERT_EQ(msg.node_entities_info_seq.size(), expected.node_entities_info_seq.size());
+  for (size_t i = 0; i < expected.node_entities_info_seq.size(); i++) {
+    const auto & node_info = msg.node_entities_info_seq[i];
+    const auto & node_info_expected = expected.node_entities_info_seq[i];
+    EXPECT_EQ(node_info.node_namespace, node_info_expected.namespace_);
+    EXPECT_EQ(node_info.node_name, node_info_expected.name);
+    auto & readers_gids = node_info.reader_gid_seq;
+    auto & expected_readers_gids = node_info_expected.reader_gids;
+    ASSERT_EQ(readers_gids.size(), expected_readers_gids.size());
+    for (size_t j = 0; j < readers_gids.size(); j++) {
+      EXPECT_EQ(readers_gids[j], gid_msg_from_string(expected_readers_gids[j]));
+    }
+    auto & writers_gids = node_info.writer_gid_seq;
+    auto & expected_writers_gids = node_info_expected.writer_gids;
+    ASSERT_EQ(writers_gids.size(), expected_writers_gids.size());
+    for (size_t j = 0; j < writers_gids.size(); j++) {
+      EXPECT_EQ(writers_gids[j], gid_msg_from_string(expected_writers_gids[j]));
+    }
+  }
+}
+
+struct EntityAssociations
+{
+  std::string gid;
+  bool is_reader;
+  std::string participant_gid;
+  std::string namespace_;
+  std::string name;
+};
+
+using EntitiesAssociations =
+  std::vector<EntityAssociations>;
```

Comment:
@ivanpauno nit: consider simply using `std::vector<EntityAssociations>`, `EntitiesAssociations` reads almost like `EntityAssociations`

Reply:
Done in https://github.com/ros2/rmw_dds_common/pull/4/commits/6247fc72227f2f483682673f714e381c3213a936.